File: modlab/ipc/backend_process.py
from PySide6.QtCore import QObject, QProcess, Signal


class BackendProcess(QObject):
    """Runs one backend subprocess per operation and parses ModLab IPC lines."""

    status_line = Signal(str, str)
    progress_update = Signal(str, int)
    operation_done = Signal(str)
    operation_error = Signal(str, str)
    raw_output = Signal(str)

    def __init__(self, python_path: str, script_path: str, parent=None):
        super().__init__(parent)
        self.python_path = python_path
        self.script_path = script_path
        self.process: QProcess | None = None
        self.current_pack = ""

    def is_busy(self) -> bool:
        return self.process is not None and self.process.state() != QProcess.NotRunning

    def run_pack(self, pack_key: str) -> None:
        self.current_pack = pack_key
        self._start_process(["--action", "run", "--pack", pack_key])

    def install_pack(self, pack_key: str) -> None:
        self.current_pack = pack_key
        self._start_process(["--action", "install", "--pack", pack_key])

    def delete_pack(self, pack_key: str) -> None:
        self.current_pack = pack_key
        self._start_process(["--action", "delete", "--pack", pack_key])

    def make_pack(self, name: str, version: str, loader: str) -> None:
        self.current_pack = name.lower()
        self._start_process(
            ["--action", "make", "--name", name, "--version", version, "--loader", loader]
        )

    def _start_process(self, args: list[str]) -> None:
        if self.process is not None:
            self.process.kill()
            self.process.deleteLater()

        self.process = QProcess(self)
        self.process.readyReadStandardOutput.connect(self._on_ready_read)
        self.process.finished.connect(self._on_finished)
        self.process.start(self.python_path, [self.script_path, *args])

    def _on_ready_read(self) -> None:
        if self.process is None:
            return
        while self.process.canReadLine():
            line = bytes(self.process.readLine()).decode("utf-8", errors="replace").strip()
            self.raw_output.emit(line)
            self._parse_line(line)

    def _parse_line(self, line: str) -> None:
        parts = line.split(":")
        if len(parts) < 2:
            return

        tag = parts[0].upper()
        pack = parts[1]
        if tag == "STATUS" and len(parts) >= 3:
            self.status_line.emit(pack, ":".join(parts[2:]))
        elif tag == "PROGRESS" and len(parts) >= 3:
            try:
                self.progress_update.emit(pack, int(parts[2]))
            except ValueError:
                pass
        elif tag == "DONE":
            self.operation_done.emit(pack)
        elif tag == "ERROR" and len(parts) >= 3:
            self.operation_error.emit(pack, ":".join(parts[2:]))

    def _on_finished(self, exit_code: int, _exit_status: QProcess.ExitStatus) -> None:
        if exit_code != 0 and self.current_pack:
            self.operation_error.emit(self.current_pack, f"Process exited with code {exit_code}")
        if self.process is not None:
            self.process.deleteLater()
            self.process = None

File: modlab-cpp-client/modlab/src/ipc/BackendProcess.cpp
#include "BackendProcess.h"
#include <QStringList>

BackendProcess::BackendProcess(const QString& pythonPath,
                               const QString& scriptPath,
                               QObject* parent)
    : QObject(parent)
    , m_python(pythonPath)
    , m_script(scriptPath)
{}

// ---- Public action methods ----

void BackendProcess::runPack(const QString& packKey) {
    m_currentPack = packKey;
    startProcess({"--action", "run", "--pack", packKey});
}

void BackendProcess::installPack(const QString& packKey) {
    m_currentPack = packKey;
    startProcess({"--action", "install", "--pack", packKey});
}

void BackendProcess::deletePack(const QString& packKey) {
    m_currentPack = packKey;
    startProcess({"--action", "delete", "--pack", packKey});
}

void BackendProcess::makePack(const QString& name,
                              const QString& version,
                              const QString& loader) {
    m_currentPack = name.toLower();
    startProcess({
        "--action", "make",
        "--name",    name,
        "--version", version,
        "--loader",  loader
    });
}

// ---- Private helpers ----

void BackendProcess::startProcess(const QStringList& args) {
    // Kill any previous process before starting a new one
    if (m_process) {
        m_process->kill();
        m_process->deleteLater();
    }

    m_process = new QProcess(this);
    connect(m_process, &QProcess::readyReadStandardOutput,
            this, &BackendProcess::onReadyRead);
    connect(m_process,
            QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
            this, &BackendProcess::onFinished);

    QStringList fullArgs = {m_script};
    fullArgs += args;
    m_process->start(m_python, fullArgs);
}

void BackendProcess::onReadyRead() {
    while (m_process->canReadLine()) {
        QString line = QString::fromUtf8(m_process->readLine()).trimmed();
        emit rawOutput(line);
        parseLine(line);
    }
}

void BackendProcess::parseLine(const QString& line) {
    // Expected formats (colon-delimited):
    //   STATUS:<pack>:<msg>
    //   PROGRESS:<pack>:<0-100>
    //   DONE:<pack>
    //   ERROR:<pack>:<msg>
    QStringList parts = line.split(':');
    if (parts.size() < 2) return;

    const QString tag  = parts[0].toUpper();
    const QString pack = parts[1];

    if (tag == "STATUS" && parts.size() >= 3) {
        emit statusLine(pack, parts.mid(2).join(':'));
    } else if (tag == "PROGRESS" && parts.size() >= 3) {
        bool ok;
        int pct = parts[2].toInt(&ok);
        if (ok) emit progressUpdate(pack, pct);
    } else if (tag == "DONE") {
        emit operationDone(pack);
    } else if (tag == "ERROR" && parts.size() >= 3) {
        emit operationError(pack, parts.mid(2).join(':'));
    }
}

void BackendProcess::onFinished(int exitCode, QProcess::ExitStatus) {
    if (exitCode != 0 && !m_currentPack.isEmpty())
        emit operationError(m_currentPack,
            QString("Process exited with code %1").arg(exitCode));
    m_process->deleteLater();
    m_process = nullptr;
}
